### PR TITLE
Display single waypoint

### PR DESCRIPTION
**Pull request by hayata-suenaga**
cc: @neil-marcellini 

### Details

This PR makes it possible to display single or multiple waypoints on the map.

### Fixed Issues

$ https://github.com/Expensify/App/issues/22706
PROPOSAL: N/A

#### First, we gonna test "flying" behavior (see the video below) when a single waypoint is provided.

1. Open NewDot on any platform
2. Go to any chat room where you can request money. Click the "+" icon next to the chat message text field. Click "Request money option from the popup menu.
3. Click "Distance" in the tab menu. Check a map of San Francisco is displayed (it might a lot time to render the map and you might see a placeholder component for a while. This is a known issue, and a fix is under work).
4. Click "Start" from the list of waypoints.
    Screenshot 2023-08-21 at 1 25 50 PM

5. On the text field in the next screen, type `88 Kearny Street.` Select the first option (the address should be San Francisco)
    Screenshot 2023-08-21 at 1 26 47 PM

6. Confirm that map "flies" and zooms into a different part of San Francisco. Refer to the following video.
    <details>
    <summary>video</summary>

    https://github.com/Expensify/App/assets/98560306/817ea994-0320-4a28-8436-b28fbe393a9b

    </details>

7. Select "Finish" from the list of waypoints.
8. On the text field in the next screen, type "Golden Gate Bridge". Click the first option that appears.
    Screenshot 2023-08-21 at 1 10 26 PM

9. Confirm that the map pans out to fit both the initial starting point and the finish point (Golden Gate Park). Refer to the following video.
    <details>
    <summary>Video</summary>

    Uploading Screen Recording 2023-08-21 at 1.28.53 PM.mov…

    </details>

10. Click the "Add stop" button. Confirm that the marker component changes. Refer to the following screenshots (the left one is "before" and the right one is "after."
    Screenshot 2023-08-21 at 1 13 22 PM
    Screenshot 2023-08-21 at 1 12 42 PM

11. Click "Finish" and type "Lands End Trail". Click the first option.
    Screenshot 2023-08-21 at 1 21 48 PM

12. Verify that the map pans out more to fit all three points/markers.
    <details>
    <summary>Video</summary>

    Uploading Screen Recording 2023-08-21 at 1.31.23 PM.mov…

    </details>

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/98560306/2dce00c2-2d0c-4dad-a9c4-ebbaebdbc60e

web multiple

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/98560306/87e7ebd2-d75d-4379-96d6-85f30c323e75

android multiple

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/98560306/e9909a53-6f0a-497c-91bb-296a25810970

safari multiple

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/98560306/f52069c8-d49c-4829-96b9-555ba8f5fb4e

desktop multiple

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/98560306/a9758474-b0d6-4169-8cc2-51c284d1a16c

Screenshot 2023-08-21 at 9 08 01 AM

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@allroundexperts I haven't tested these on Android and iOS native apps. I'd really appreciate it if you could test this PR on these platforms 🙇

File: README.md
Comment (hayata-suenaga) [reply README.md:60]: nice idea 💡 but the script is interactive (the user has to create a Mapbox account and then generate a token)

are you suggesting that we should check if the token is already set in the user' computer and if not prompt the user to generate token?

Comment (hayata-suenaga) [reply README.md:60]: that has been discussed but there is no good way to share a secret token right now. We gonna consider that in the future (we know how bad this is in terms of DX and understand this might potentially be a barrier to new contributors). 😓

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:134]: ahh so sorry I forgot to mention that this PR was actually spanned from [another PR](https://github.com/Expensify/App/pull/24306)

A lot of changes on this PR are actually from that PR and most of the issues are addressed there

I gonna provide you a link? to the portion of the code that is specific to this PR. Let me do that after I get the lunch and do other issues in about 2 hours

again, so sorry that I forgot to mention that

**Comment by hayata-suenaga** [comment]
Found a bug

when changing the window size while there is a single waypoint, the map is reset and the map focuses on a wrong location
https://github.com/Expensify/App/assets/98560306/b1ce68fd-5490-46d3-ab65-abfdc91e725d

**Comment by hayata-suenaga** [comment]
This bug stems from how the Right Side Panel (I don't know if this is the right name) is resized when a breakpoint is reached (if you shrink the browser window, the size of Right Side Panel changes to cover the entire page).

I asked a question regarding this in the #epensify-open-source channel Slack channel.

While this question being asked, I gonna implement an alternative solution where we listen to the window size change and adjust the map accordingly.

**Comment by hayata-suenaga** [comment]
@allroundexperts the PR is ready (except that I have to change `react-native-x-maps` to use version number after [this PR](https://github.com/Expensify/react-native-x-maps/pull/23) is merged and the version is bumped).

I'm still having trouble with local Android build, so if you can test this PR on Android emulator, I'd appreciate it. I haven't taken screenshots for iOS native (but I tested). I can take screenshots for iOS native tomorrow morning.

**Comment by hayata-suenaga** [comment]
@allroundexperts nop 😅 updated the title 🙇

**Comment by hayata-suenaga** [comment]
> I haven't tested these on Android and iOS native apps. I'd really appreciate it if you could test this PR on these platforms 🙇

I tested on iOS and added screenshot in the OP. I still can't build Android, so I gonna rely on your test and screenshots for it @allroundexperts 🙇

File: src/components/DistanceRequest.js
Comment (neil-marcellini) [inline src/components/DistanceRequest.js:85]: NAB: We could simplify to `if (!waypoint.lng || !waypoint.lat)`

File: src/components/DistanceRequest.js
Comment (neil-marcellini) [inline src/components/DistanceRequest.js:105]: I think we should use theme.icon since it needs to match the backend and that's what I'm using there. But, let's check with the design team.

File: src/components/DistanceRequest.js
Comment (neil-marcellini) [inline src/components/DistanceRequest.js:83]: NAB: Let's use a more descriptive name since this is now an object of `waypointMarkers`.

**Review by neil-marcellini — APPROVED** [review]
Looking good pending C+ testing.

**Comment by neil-marcellini** [comment]
Let's also get a review from @Expensify/design. It seems like the design label doesn't auto assign here.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Oh also @hayata-suenaga, can we please start using the Expensify custom map style? The style URL can be built using username `expensify` and style_id `cllcoiqds00cs01r80kp34tmq`. Here are the [docs for that](https://docs.mapbox.com/help/glossary/style-url/).

**Comment by shawnborton** [comment]
Oh interesting, let me check in on the label not working...

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:85]: it's possible that either longitude and latitude is `0` 🤔

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:105]: `them.icon` is this color. Is this what's we're looking for? cc: @shawnborton 

Screenshot 2023-08-21 at 1 36 59 PM

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:83]: changed 👍

**Comment by hayata-suenaga** [comment]
@allroundexperts just a heads-up. We modified the test step for this PR quite drastically [here](https://github.com/Expensify/App/pull/25161#issue-1854256016). I apologize if you already started testing, and I'd appreciate it you could follow the new test steps 🙇 Please let me know if you have any questions

**Comment by hayata-suenaga** [comment]
> Oh interesting, let me check in on the label not working...

Please let me know if the icons/markers look okay on the map. Most of the screenshots/vidoes are in [the original post](https://github.com/Expensify/App/pull/25161#issue-1854256016) of this PR.

By the way, the mechanism we use to display markers in this PR is different from what are being discussed in [this Slack thread](https://expensify.slack.com/archives/C05DWUDHVK7/p1692583625196909). Changes you make on Mapbox for markers won't affect how markers look in this PR.

**Comment by shawnborton** [comment]
Ah, got it. I think what you have here is good, noting that we'll likely swap out images after this is merged.

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:105]: I think the color might change after the design discussion happening [in this thread](https://expensify.slack.com/archives/C05DWUDHVK7/p1692583625196909). So for now, we can just go with `theme.icon` and change if necessary after the merge.

[We probably have to change the icons themselves](https://github.com/Expensify/App/pull/25161#issuecomment-1687032963) anyway.

**Comment by hayata-suenaga** [comment]
### Memo

If this PR is merged before the design discussion is included for the marker style, we gonna open a new PR to replace marker icons.

File: src/components/DistanceRequest.js
Comment (neil-marcellini) [reply src/components/DistanceRequest.js]: Oh true good point

**Comment by hayata-suenaga** [comment]
@allroundexperts thank you for thorough testing 🚀 

> Panning the map, slides the whole navigator.

This issue is known and will be handled in [this separate issue](https://github.com/Expensify/App/issues/25422).

> Map isn't responsive when the screen size changes.

This issue is also known and this is [a design choice we made in this PR](https://github.com/Expensify/App/pull/24306#issuecomment-1681324194).

**Comment by hayata-suenaga** [comment]
Both the map itself and map placeholder component is displayed properly on my emulator.
Screenshot 2023-08-21 at 3 24 32 PM
Screenshot 2023-08-21 at 3 27 04 PM


However, I realized `notFound.goBackHome` is displayed as a text button somehow. I gonna look into this but do you have any idea why this is happening @allroundexperts 😭 ?

**Comment by hayata-suenaga** [comment]
> @hayata-suenaga This is happening because you're not passing in the `link` prop [here](https://github.com/Expensify/App/blob/40eecdb61a851807bdb4808579237ad747823770/src/components/DistanceRequest.js#L211-L215).
> 
> By default its value is [this](https://github.com/Expensify/App/blob/40eecdb61a851807bdb4808579237ad747823770/src/components/BlockingViews/BlockingView.js#L46-L47) but we're just using the key as it is without translating it.
> 
> You can also set `shouldShowLink` as `false` to hide the link.

ah didn't know that. very interesting API 🤔 anyway, I gonna set `shouldShowLink` to `false`

**Comment by hayata-suenaga** [comment]
@allroundexperts are you still having trouble with android and ios browsers?

**Comment by hayata-suenaga** [comment]
> But for iOS safari, I think the problem is real.

@allroundexperts I really appreciate you found this issue. But this wasn't caused by this PR so we can handle the fix in a separate issue.

If there is no other issue, could you complete the checklist please? 🙇

**Review by allroundexperts — APPROVED** [review]
Testing well!

**Comment by hayata-suenaga** [comment]
The scrolling behavior of the waypoint list works fine on iOS
https://github.com/Expensify/App/assets/98560306/9770e573-8cc5-4a4f-9549-d5880c14d473

I gonna test on SE

**Comment by hayata-suenaga** [comment]
I feel like I'm missing something here 😕 the waypoint list is scrollable on SE, too

https://github.com/Expensify/App/assets/98560306/7bafc2c5-6753-48df-847a-bf7c24f9608e

**Comment by hayata-suenaga** [comment]
> Okay. It turns out that if you try to scroll by pressing on the Stop text, it does not work. However, panning on anything above that does work.

okay on the actual physical device, I think it would be much easier to scroll, but thank you for the investigation 🥇

**Review by neil-marcellini — APPROVED** [review]
Looks great, awesome work!

**Comment by thienlnam** [comment]
@allroundexperts Found that auto-panning doesn't seem to work if you add the following addresses on iOS

Start: 584 Castro St, San Francisco
End: 88 Kearney St, South San Francisco, CA 94080, USA

https://expensify.slack.com/archives/C02NK2DQWUX/p1692731847794059?thread_ts=1692665981.227269&cid=C02NK2DQWUX

**Comment by hayata-suenaga** [comment]
@thienlnam did you merge main?

File: src/components/DistanceRequest.js
Comment (s77rt) [inline src/components/DistanceRequest.js:208]: After adding waypoints, we should have changed the default location from SF to the first value in waypoints to keep animations consist (coming from https://github.com/Expensify/App/issues/29471)

Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:208]: ah that's a good point. Thank for letting us know!

**Pull request merged**
